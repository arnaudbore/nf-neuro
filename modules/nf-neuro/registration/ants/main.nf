
process REGISTRATION_ANTS {
    tag "$meta.id"
    label 'process_medium'

    container "scilus/scilus:2.2.0"

    input:
        tuple val(meta), path(fixed_image), path(moving_image), path(mask) //** optional, input = [] **//

    output:
        tuple val(meta), path("*_warped.nii.gz")                            , emit: image_warped
        tuple val(meta), path("*_forward1_affine.mat")                     , emit: forward_affine, optional: true
        tuple val(meta), path("*_forward0_warp.nii.gz")                    , emit: forward_warp, optional: true
        tuple val(meta), path("*_backward1_warp.nii.gz")                   , emit: backward_warp, optional: true
        tuple val(meta), path("*_backward0_affine.mat")                    , emit: backward_affine, optional: true
        tuple val(meta), path("*_forward*.{nii.gz,mat}", arity: '1..2')    , emit: forward_image_transform
        tuple val(meta), path("*_backward*.{nii.gz,mat}", arity: '1..2')   , emit: backward_image_transform
        tuple val(meta), path("*_backward*.{nii.gz,mat}", arity: '1..2')   , emit: forward_tractogram_transform
        tuple val(meta), path("*_forward*.{nii.gz,mat}", arity: '1..2')    , emit: backward_tractogram_transform
        tuple val(meta), path("*_registration_ants_mqc.gif")                , emit: mqc, optional: true
        path "versions.yml"                                                 , emit: versions

    when:
        task.ext.when == null || task.ext.when

    script:
    def initialization_types = ["geometric center": 0, "intensities": 1, "origin": 2]
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def suffix_qc = task.ext.suffix_qc ?: ""
    def ants = task.ext.quick ? "antsRegistrationSyNQuick.sh" : "antsRegistrationSyN.sh"
    def dimension = "-d ${task.ext.dimension ?: 3}"
    def transform = task.ext.transform ?: "s"
    def seed = " -e ${task.ext.random_seed ?: 1234}"
    def run_qc = task.ext.run_qc as Boolean || false

    args += " -n $task.cpus"
    if ( mask ) args += " -x $mask"
    if ( task.ext.initial_transform ) args += " -i [$fixed_image,$moving_image,${initialization_types[task.ext.initial_transform]}]"
    if ( task.ext.histogram_bins ) args += " -r $task.ext.histogram_bins"
    if ( task.ext.spline_distance ) args += " -s $task.ext.spline_distance"
    if ( task.ext.gradient_step ) args += " -g $task.ext.gradient_step"
    if ( task.ext.precision ) args += " -p $task.ext.precision"
    if ( task.ext.histogram_matching ) args += " -j $task.ext.histogram_matching"
    if ( task.ext.repro_mode ) args += " -y $task.ext.repro_mode"
    if ( task.ext.collapse_output ) args += " -z $task.ext.collapse_output"

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=$task.cpus
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    $ants $dimension -f $fixed_image -m $moving_image -o output -t $transform $args $seed

    moving_id=\$(basename $moving_image .nii.gz)
    moving_id=\${moving_id#${meta.id}_*}

    mv outputWarped.nii.gz ${prefix}_\${moving_id}_warped.nii.gz

    if [ $transform != "bo" ] && [ $transform != "so" ]; then
        mv output0GenericAffine.mat ${prefix}_forward1_affine.mat
    fi

    if [ $transform != "t" ] && [ $transform != "r" ] && [ $transform != "a" ]; then
        mv output1InverseWarp.nii.gz ${prefix}_backward1_warp.nii.gz
        mv output1Warp.nii.gz ${prefix}_forward0_warp.nii.gz
    fi

    antsApplyTransforms -d 3 -t [${prefix}_forward1_affine.mat,1] \
        -o Linear[${prefix}_backward0_affine.mat]

    ### ** QC ** ###
    if $run_qc; then
        mv $fixed_image fixed_image.nii.gz
        extract_dim=\$(mrinfo fixed_image.nii.gz -size)
        read sagittal_dim coronal_dim axial_dim <<< "\${extract_dim}"

        # Get the middle slice
        coronal_dim=\$((\$coronal_dim / 2))
        axial_dim=\$((\$axial_dim / 2))
        sagittal_dim=\$((\$sagittal_dim / 2))

        # Get fixed ID, moving ID already computed
        fixed_id=\$(basename $fixed_image .nii.gz)
        fixed_id=\${fixed_id#${meta.id}_*}

        # Set viz params.
        viz_params="--display_slice_number --display_lr --size 256 256"
        # Iterate over images.
        for image in fixed_image warped; do
            mrconvert *\${image}.nii.gz *\${image}_viz.nii.gz -stride -1,2,3
            scil_viz_volume_screenshot *\${image}_viz.nii.gz \${image}_coronal.png \
                --slices \$coronal_dim --axis coronal \$viz_params
            scil_viz_volume_screenshot *\${image}_viz.nii.gz \${image}_sagittal.png \
                --slices \$sagittal_dim --axis sagittal \$viz_params
            scil_viz_volume_screenshot *\${image}_viz.nii.gz \${image}_axial.png \
                --slices \$axial_dim --axis axial \$viz_params

            if [ \$image != fixed_image ]; then
                title="Warped \${moving_id^^}"
            else
                title="Reference \${fixed_id^^}"
            fi

            convert +append \${image}_coronal*.png \${image}_axial*.png \
                \${image}_sagittal*.png \${image}_mosaic.png
            convert -annotate +20+230 "\${title}" -fill white -pointsize 30 \
                \${image}_mosaic.png \${image}_mosaic.png

            # Clean up.
            rm \${image}_coronal*.png \${image}_sagittal*.png \${image}_axial*.png
        done

        # Create GIF.
        convert -delay 10 -loop 0 -morph 10 \
            warped_mosaic.png fixed_image_mosaic.png warped_mosaic.png \
            ${prefix}_${suffix_qc}_registration_ants_mqc.gif

        # Clean up.
        rm *_mosaic.png
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ants: \$(antsRegistration --version | grep "Version" | sed -E 's/.*: v?([0-9.a-zA-Z-]+).*/\\1/')
        imagemagick: \$(convert -version | grep "Version:" | sed -E 's/.*ImageMagick ([0-9.-]+).*/\\1/')
        mrtrix: \$(mrinfo -version 2>&1 | grep "== mrinfo" | sed -E 's/== mrinfo ([0-9.]+).*/\\1/')
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def suffix_qc = task.ext.suffix_qc ?: ""
    def run_qc = task.ext.run_qc as Boolean || false

    """
    set +e
    function handle_code () {
        local code=\$?
        ignore=( 1 )
        [[ " \${ignore[@]} " =~ " \$code " ]] || exit \$code
    }

    # Local trap to ignore awaited non-zero exit codes
    {
        trap 'handle_code' ERR
        antsRegistrationSyNQuick.sh -h
    }

    antsApplyTransforms -h
    convert -help .
    scil_viz_volume_screenshot -h

    touch ${prefix}_t1_warped.nii.gz
    touch ${prefix}_forward1_affine.mat
    touch ${prefix}_forward0_warp.nii.gz
    touch ${prefix}_backward1_warp.nii.gz
    touch ${prefix}_backward0_affine.mat

    if $run_qc; then
        touch ${prefix}_${suffix_qc}_registration_ants_mqc.gif
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ants: \$(antsRegistration --version | grep "Version" | sed -E 's/.*: v?([0-9.a-zA-Z-]+).*/\\1/')
        imagemagick: \$(convert -version | grep "Version:" | sed -E 's/.*ImageMagick ([0-9.-]+).*/\\1/')
        mrtrix: \$(mrinfo -version 2>&1 | grep "== mrinfo" | sed -E 's/== mrinfo ([0-9.]+).*/\\1/')
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """
}
