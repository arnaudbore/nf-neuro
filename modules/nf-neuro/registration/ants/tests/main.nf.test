nextflow_process {

    name "Test Process REGISTRATION_ANTS"
    script "../main.nf"
    process "REGISTRATION_ANTS"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfneuro"
    tag "registration"
    tag "registration/ants"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "T1w.zip", "others.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("registration - ants - SyN") {
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        T1w: it.simpleName == "T1w"
                        moving: it.simpleName == "others"
                    }
                ch_T1w = ch_split_test_data.T1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]
                }
                ch_moving = ch_split_test_data.moving.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1.nii.gz")
                    ]
                }
                ch_T1w_mask = ch_split_test_data.T1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w_mask.nii.gz")
                    ]
                }
                input[0] = ch_T1w
                    .join(ch_moving)
                    .join(ch_T1w_mask)
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("registration - ants - SyN quick") {
        config "./nextflow_quick.config"
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        T1w: it.simpleName == "T1w"
                        moving: it.simpleName == "others"
                    }
                ch_T1w = ch_split_test_data.T1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]
                }
                ch_moving = ch_split_test_data.moving.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1.nii.gz")
                    ]
                }
                ch_T1w_mask = ch_split_test_data.T1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w_mask.nii.gz")
                    ]
                }
                input[0] = ch_T1w
                    .join(ch_moving)
                    .join(ch_T1w_mask)
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("registration - ants - no warps") {
        config "./nextflow_no_warp.config"
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        T1w: it.simpleName == "T1w"
                        moving: it.simpleName == "others"
                    }
                ch_T1w = ch_split_test_data.T1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]
                }
                ch_moving = ch_split_test_data.moving.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1.nii.gz")
                    ]
                }
                ch_T1w_mask = ch_split_test_data.T1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w_mask.nii.gz")
                    ]
                }
                input[0] = ch_T1w
                    .join(ch_moving)
                    .join(ch_T1w_mask)
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("registration - ants - stub") {
        tag "stub"
        options "-stub-run"
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        T1w: it.simpleName == "T1w"
                        moving: it.simpleName == "others"
                    }
                ch_T1w = ch_split_test_data.T1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]
                }
                ch_moving = ch_split_test_data.moving.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1.nii.gz")
                    ]
                }
                ch_T1w_mask = ch_split_test_data.T1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w_mask.nii.gz")
                    ]
                }
                input[0] = ch_T1w
                    .join(ch_moving)
                    .join(ch_T1w_mask)
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match() }
            )
        }
    }
}
