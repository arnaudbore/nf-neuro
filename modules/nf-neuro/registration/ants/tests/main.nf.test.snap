{
    "registration - ants - stub": {
        "content": [
            [
                "versions.yml:md5,78610891ed0adf1a5fe5f0de034d7555"
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-28T13:00:52.757761907"
    },
    "registration - ants - SyN quick": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_warped.nii.gz:md5,5534f8cb21fcc44f577d520e6083bb83"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward1_affine.mat:md5,dde4de8a1ca4ff1647e55dc1fece402c"
                    ]
                ],
                "10": [
                    "versions.yml:md5,78610891ed0adf1a5fe5f0de034d7555"
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward0_warp.nii.gz:md5,96b2d18c7537db76366e4557b44f24fe"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward1_warp.nii.gz:md5,28f58437d878a6ea0c81c0e780e2333f"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_affine.mat:md5,8cc894275b42e3b78efca499caf22e30"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_forward0_warp.nii.gz:md5,96b2d18c7537db76366e4557b44f24fe",
                            "test_forward1_affine.mat:md5,dde4de8a1ca4ff1647e55dc1fece402c"
                        ]
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_backward0_affine.mat:md5,8cc894275b42e3b78efca499caf22e30",
                            "test_backward1_warp.nii.gz:md5,28f58437d878a6ea0c81c0e780e2333f"
                        ]
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_backward0_affine.mat:md5,8cc894275b42e3b78efca499caf22e30",
                            "test_backward1_warp.nii.gz:md5,28f58437d878a6ea0c81c0e780e2333f"
                        ]
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_forward0_warp.nii.gz:md5,96b2d18c7537db76366e4557b44f24fe",
                            "test_forward1_affine.mat:md5,dde4de8a1ca4ff1647e55dc1fece402c"
                        ]
                    ]
                ],
                "9": [
                    [
                        {
                            "id": "test"
                        },
                        "test_T1_to_T1_slab_registration_ants_mqc.gif:md5,e5730d99d11edfc9802776c782edbf17"
                    ]
                ],
                "backward_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_affine.mat:md5,8cc894275b42e3b78efca499caf22e30"
                    ]
                ],
                "backward_image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_backward0_affine.mat:md5,8cc894275b42e3b78efca499caf22e30",
                            "test_backward1_warp.nii.gz:md5,28f58437d878a6ea0c81c0e780e2333f"
                        ]
                    ]
                ],
                "backward_tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_forward0_warp.nii.gz:md5,96b2d18c7537db76366e4557b44f24fe",
                            "test_forward1_affine.mat:md5,dde4de8a1ca4ff1647e55dc1fece402c"
                        ]
                    ]
                ],
                "backward_warp": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward1_warp.nii.gz:md5,28f58437d878a6ea0c81c0e780e2333f"
                    ]
                ],
                "forward_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward1_affine.mat:md5,dde4de8a1ca4ff1647e55dc1fece402c"
                    ]
                ],
                "forward_image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_forward0_warp.nii.gz:md5,96b2d18c7537db76366e4557b44f24fe",
                            "test_forward1_affine.mat:md5,dde4de8a1ca4ff1647e55dc1fece402c"
                        ]
                    ]
                ],
                "forward_tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_backward0_affine.mat:md5,8cc894275b42e3b78efca499caf22e30",
                            "test_backward1_warp.nii.gz:md5,28f58437d878a6ea0c81c0e780e2333f"
                        ]
                    ]
                ],
                "forward_warp": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward0_warp.nii.gz:md5,96b2d18c7537db76366e4557b44f24fe"
                    ]
                ],
                "image_warped": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_warped.nii.gz:md5,5534f8cb21fcc44f577d520e6083bb83"
                    ]
                ],
                "mqc": [
                    [
                        {
                            "id": "test"
                        },
                        "test_T1_to_T1_slab_registration_ants_mqc.gif:md5,e5730d99d11edfc9802776c782edbf17"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,78610891ed0adf1a5fe5f0de034d7555"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-28T13:00:13.19463937"
    },
    "registration - ants - SyN": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_warped.nii.gz:md5,8d9c289924c4d2c8edab3feae669d1c3"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward1_affine.mat:md5,16a42a74c35c9fda7786250b333bef86"
                    ]
                ],
                "10": [
                    "versions.yml:md5,78610891ed0adf1a5fe5f0de034d7555"
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward0_warp.nii.gz:md5,9a3b969b74ac82ab0940b8b94677c7b9"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward1_warp.nii.gz:md5,3f2d23cce4fbcf970099eecd629d4019"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_affine.mat:md5,317e067757768c8e75de9d12f84563c9"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_forward0_warp.nii.gz:md5,9a3b969b74ac82ab0940b8b94677c7b9",
                            "test_forward1_affine.mat:md5,16a42a74c35c9fda7786250b333bef86"
                        ]
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_backward0_affine.mat:md5,317e067757768c8e75de9d12f84563c9",
                            "test_backward1_warp.nii.gz:md5,3f2d23cce4fbcf970099eecd629d4019"
                        ]
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_backward0_affine.mat:md5,317e067757768c8e75de9d12f84563c9",
                            "test_backward1_warp.nii.gz:md5,3f2d23cce4fbcf970099eecd629d4019"
                        ]
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_forward0_warp.nii.gz:md5,9a3b969b74ac82ab0940b8b94677c7b9",
                            "test_forward1_affine.mat:md5,16a42a74c35c9fda7786250b333bef86"
                        ]
                    ]
                ],
                "9": [
                    
                ],
                "backward_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_affine.mat:md5,317e067757768c8e75de9d12f84563c9"
                    ]
                ],
                "backward_image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_backward0_affine.mat:md5,317e067757768c8e75de9d12f84563c9",
                            "test_backward1_warp.nii.gz:md5,3f2d23cce4fbcf970099eecd629d4019"
                        ]
                    ]
                ],
                "backward_tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_forward0_warp.nii.gz:md5,9a3b969b74ac82ab0940b8b94677c7b9",
                            "test_forward1_affine.mat:md5,16a42a74c35c9fda7786250b333bef86"
                        ]
                    ]
                ],
                "backward_warp": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward1_warp.nii.gz:md5,3f2d23cce4fbcf970099eecd629d4019"
                    ]
                ],
                "forward_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward1_affine.mat:md5,16a42a74c35c9fda7786250b333bef86"
                    ]
                ],
                "forward_image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_forward0_warp.nii.gz:md5,9a3b969b74ac82ab0940b8b94677c7b9",
                            "test_forward1_affine.mat:md5,16a42a74c35c9fda7786250b333bef86"
                        ]
                    ]
                ],
                "forward_tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_backward0_affine.mat:md5,317e067757768c8e75de9d12f84563c9",
                            "test_backward1_warp.nii.gz:md5,3f2d23cce4fbcf970099eecd629d4019"
                        ]
                    ]
                ],
                "forward_warp": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward0_warp.nii.gz:md5,9a3b969b74ac82ab0940b8b94677c7b9"
                    ]
                ],
                "image_warped": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_warped.nii.gz:md5,8d9c289924c4d2c8edab3feae669d1c3"
                    ]
                ],
                "mqc": [
                    
                ],
                "versions": [
                    "versions.yml:md5,78610891ed0adf1a5fe5f0de034d7555"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-28T12:58:47.595933738"
    },
    "registration - ants - no warps": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_warped.nii.gz:md5,c13330e3f71c57f2517dcd6f59f64ff1"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward1_affine.mat:md5,7387bc8f93dd29d2a0aadb9d5dc06d5b"
                    ]
                ],
                "10": [
                    "versions.yml:md5,78610891ed0adf1a5fe5f0de034d7555"
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_affine.mat:md5,733babe33e1656d0f85064caf24d6f82"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_forward1_affine.mat:md5,7387bc8f93dd29d2a0aadb9d5dc06d5b"
                        ]
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_backward0_affine.mat:md5,733babe33e1656d0f85064caf24d6f82"
                        ]
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_backward0_affine.mat:md5,733babe33e1656d0f85064caf24d6f82"
                        ]
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_forward1_affine.mat:md5,7387bc8f93dd29d2a0aadb9d5dc06d5b"
                        ]
                    ]
                ],
                "9": [
                    
                ],
                "backward_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_affine.mat:md5,733babe33e1656d0f85064caf24d6f82"
                    ]
                ],
                "backward_image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_backward0_affine.mat:md5,733babe33e1656d0f85064caf24d6f82"
                        ]
                    ]
                ],
                "backward_tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_forward1_affine.mat:md5,7387bc8f93dd29d2a0aadb9d5dc06d5b"
                        ]
                    ]
                ],
                "backward_warp": [
                    
                ],
                "forward_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward1_affine.mat:md5,7387bc8f93dd29d2a0aadb9d5dc06d5b"
                    ]
                ],
                "forward_image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_forward1_affine.mat:md5,7387bc8f93dd29d2a0aadb9d5dc06d5b"
                        ]
                    ]
                ],
                "forward_tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_backward0_affine.mat:md5,733babe33e1656d0f85064caf24d6f82"
                        ]
                    ]
                ],
                "forward_warp": [
                    
                ],
                "image_warped": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_warped.nii.gz:md5,c13330e3f71c57f2517dcd6f59f64ff1"
                    ]
                ],
                "mqc": [
                    
                ],
                "versions": [
                    "versions.yml:md5,78610891ed0adf1a5fe5f0de034d7555"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-28T13:00:29.996790917"
    }
}